Add runningStatus condition support.

yoshisatoyanagisawa opened the pull request:
The ServiceWorker static routing API has a condition on a ServiceWorker running status. This is a specification update to support that.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/5.html" title="Last updated on Jan 18, 2024, 2:37 AM UTC (8a6887b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/5/a44f9a4...8a6887b.html" title="Last updated on Jan 18, 2024, 2:37 AM UTC (8a6887b)">Diff</a>

yoshisatoyanagisawa committed:
Add runningStatus condition support.

yoshisatoyanagisawa commented:
@sisidovski @domenic Can I ask you to review the specification update on the running status?
Thank you.

sisidovski commented on docs/index.bs:
```diff
@@ -1563,8 +1563,10 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
 
       dictionary RouterCondition {
         URLPatternCompatible urlPattern;
+        RunningStatus runningStatus;
```
Adding the new entry to the dictionary. I understood this means `RouterCondition` can have multiple conditions like following:
```js
condition: {
  urlPattern: "/test",
  runningStatus: "running"
}
```

And it makes sense that the SW treats the request as a matched request when all of conditions are satisfied.

yoshisatoyanagisawa replied:
I have added notes to explain that.

sisidovski commented on docs/index.bs:
```diff
@@ -3259,10 +3266,14 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. If running [=match=] with |pattern| and |request|'s [=request/URL=] returns null, [=continue=].
+          1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/runningStatus}}"] [=map/exists=], then:
+              1. Let |runningStatus| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/runningStatus}}"].
+              1. If |runningStatus| is {{RunningStatus/"running"}}, and |serviceWorker| is not [=running=], continue.
```
[=continue=]

yoshisatoyanagisawa replied:
done.

sisidovski commented on docs/index.bs:
```diff
@@ -3259,10 +3266,14 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+          1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/runningStatus}}"] [=map/exists=], then:
+              1. Let |runningStatus| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/runningStatus}}"].
+              1. If |runningStatus| is {{RunningStatus/"running"}}, and |serviceWorker| is not [=running=], continue.
+              1. If |runningStatus| is {{RunningStatus/"not-running"}}, and |serviceWorker| is [=running=], continue.
```
ditto

yoshisatoyanagisawa replied:
done.

sisidovski reviewed (approved):
LGTM w/ nits

yoshisatoyanagisawa committed:
Addressed comments.

- added a note on multiple conditions.
- changed continue to [=continue=].

domenic commented on docs/index.bs:
```diff
@@ -3240,14 +3242,19 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
-      1. Return true.
+          1. Set |hasCondition| to true.
+      1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/runningStatus}}"] [=map/exists=], then:
+          1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/runningStatus}}"] is neither of  {{RunningStatus/"running"}} nor {{RunningStatus/"not-running"}}, return false.
```
I think this is prohibited by IDL. Since _rule_["`condition`"]["`runningStatus`"] is of type `RunningStatus`, it can never have a different value from those two.

yoshisatoyanagisawa replied:
sure.  I have removed the check, and directly set |hasCondition| true.

domenic commented on docs/index.bs:
```diff
@@ -3259,10 +3266,17 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
-          1. Let |pattern| be the result of running <a>Parse URL Pattern</a> algorithm passing |rawPattern| and |serviceWorker|.
-          1. If running [=match=] with |pattern| and |request|'s [=request/URL=] returns null, [=continue=].
+
+          Note: if there are multiple conditions in a rule, all conditions should be matched to return the source.
```
```suggestion
          Note: if there are multiple conditions in a rule, all conditions will be matched to return the source.
```

Try to avoid keywords like "must" or "should" inside non-normative explanatory notes.

yoshisatoyanagisawa replied:
done.

domenic reviewed (approved):
LGTM with nits

yoshisatoyanagisawa committed:
Addressed comment.

- remove checking if runningStatus is one of RunningStatus.
- use "will" instead of "should".

yoshisatoyanagisawa commented:
Thanks for the review.  I am going to merge the change.
